apps/designer/app/shared/copy-paste/use-instance-copy-paste.ts

```diff
@@ -0,0 +1,163 @@
+  onCut?: (instance: Instance) => void;
+};
+
+const createEventsHandler = () => {
+  let currentProps: Props | undefined;
+
+  const handleCopy = (event: ClipboardEvent) => {
+    if (currentProps === undefined) {
+      return;
+    }
+    const { selectedInstanceData, allowAnyTarget } = currentProps;
+
+    if (
+      event.clipboardData === null ||
+      selectedInstanceData === undefined ||
+      isInstanceClipboardEvent(event, { allowAnyTarget }) === false
+    ) {
+      return;
+    }
+
+    // must prevent default, otherwise setData() will not work
+    event.preventDefault();
+
+    event.clipboardData.setData(
+      "application/json",
+      serialize(selectedInstanceData)
+    );
+  };
+
+  const handleCut = (event: ClipboardEvent) => {
+    if (currentProps === undefined) {
+      return;
+    }
+    const { selectedInstanceData, allowAnyTarget, onCut } = currentProps;
+
+    if (
+      event.clipboardData === null ||
+      selectedInstanceData === undefined ||
+      isInstanceClipboardEvent(event, {
+        allowAnyTarget: allowAnyTarget,
+      }) === false
+    ) {
+      return;
+    }
+
+    // must prevent default, otherwise setData() will not work
+    event.preventDefault();
+
+    event.clipboardData.setData(
+      "application/json",
+      serialize(selectedInstanceData)
+    );
+
+    onCut?.(selectedInstanceData.instance);
+  };
+
+  const handlePaste = (event: ClipboardEvent) => {
+    if (currentProps === undefined) {
+      return;
+    }
+    const { onPaste, allowAnyTarget } = currentProps;
+
+    if (
+      onPaste === undefined ||
+      event.clipboardData === null ||
+      // we might want a separate predicate for paste,
+      // but for now the logic is the same
+      isInstanceClipboardEvent(event, { allowAnyTarget }) === false
+    ) {
+      return;
+    }
+
+    // this shouldn't matter, but just in case
+    event.preventDefault();
+
+    const data = deserialize(event.clipboardData.getData("application/json"));
+
+    if (data === undefined) {
+      return;
+    }
+
+    const instance = utils.tree.cloneInstance(data.instance);
+
+    const props =
+      data.props && data.props.length > 0
+        ? utils.props.cloneUserProps(data.props)
+        : undefined;
+
+    onPaste(instance, props);
+  };
+
+  return {
+    start() {
+      document.addEventListener("copy", handleCopy);
+      document.addEventListener("cut", handleCut);
+      document.addEventListener("paste", handlePaste);
+    },
+    stop() {
+      document.removeEventListener("copy", handleCopy);
+      document.removeEventListener("cut", handleCut);
+      document.removeEventListener("paste", handlePaste);
+    },
+    setProps(props: Props) {
+      currentProps = props;
+    },
+  };
+};
+
+// Everything is extrcated from hook,
+// to make it easier to remove React from canvas if we need to
+export const useInstanceCopyPaste = (props: Props): void => {
+  const eventsHandler = useMemo(() => createEventsHandler(), []);
```

Comment:
seems you could even do
```suggestion
  const eventsHandler = useMemo(createEventsHandler, []);
```